examples/lora_spi_gps.rs

```diff
@@ -0,0 +1,880 @@
+
+const CONFIG_RADIO: radio_sx127x::device::Config = radio_sx127x::device::Config {
+    modem: Modem::LoRa(CONFIG_LORA),
+    channel: Channel::LoRa(CONFIG_CH),
+    pa_config: CONFIG_PA,
+    xtal_freq: 32000000, // CHECK
+    timeout_ms: 100,
+};
+
+// setup() does all  hal/MCU specific setup and returns generic hal device for use in main code.
+
+#[cfg(feature = "stm32f0xx")] //  eg stm32f030xc
+use stm32f0xx_hal::{
+    delay::Delay,
+    pac::Peripherals,
+    pac::USART2,
+    prelude::*,
+    serial::{Rx, Serial, Tx},
+    spi::{Error, Spi},
+};
+
+#[cfg(feature = "stm32f0xx")]
+fn setup() -> (
+    Tx<USART2>,
+    Rx<USART2>,
+    impl DelayMs<u32> + Transmit<Error = sx127xError<Error, Infallible, Infallible>>,
+) {
+    let cp = cortex_m::Peripherals::take().unwrap();
+    let mut p = Peripherals::take().unwrap();
+    let mut rcc = p.RCC.configure().freeze(&mut p.FLASH);
+
+    let gpioa = p.GPIOA.split(&mut rcc);
+    let gpiob = p.GPIOB.split(&mut rcc);
+
+    //  stm32f030xc builds with gpiob..into_alternate_af4(cs) USART3 on tx pb10, rx pb11
+    //    but stm32f042  only has 2 usarts.
+    //  Both have gpioa..into_alternate_af1(cs) USART2 with tx on pa2 and rx pa3
+
+    let (tx, rx, sck, miso, mosi, _rst, pa1, pb8, pb9, pa0) =
+        cortex_m::interrupt::free(move |cs| {
+            (
+                gpioa.pa2.into_alternate_af1(cs), //tx pa2  for GPS
+                gpioa.pa3.into_alternate_af1(cs), //rx pa3  for GPS
+                gpioa.pa5.into_alternate_af0(cs), //   sck   on PA5
+                gpioa.pa6.into_alternate_af0(cs), //   miso  on PA6
+                gpioa.pa7.into_alternate_af0(cs), //   mosi  on PA7
+                //gpioa.pa1.into_push_pull_output(cs),	  //   cs            on PA1
+                gpiob.pb1.into_push_pull_output(cs), //   reset         on PB1
+                gpioa.pa1.into_push_pull_output(cs), //   CsPin	     on PA1
+                gpiob.pb8.into_floating_input(cs),   //   BusyPin  DIO0 on PB8
+                gpiob.pb9.into_floating_input(cs),   //   ReadyPin DIO1 on PB9
+                gpioa.pa0.into_push_pull_output(cs), //   ResetPin	      on PA0
+            )
+        });
+
+    let (tx, rx) = Serial::usart2(p.USART2, (tx, rx), 9600.bps(), &mut rcc).split();
+
+    let spi = Spi::spi1(p.SPI1, (sck, miso, mosi), MODE, 8.mhz(), &mut rcc);
+
+    let delay = Delay::new(cp.SYST, &rcc);
+
+    // Create lora radio instance
+
+    let lora = Sx127x::spi(
+        spi.compat(),
+        pa1.compat(),
+        pb8.compat(),
+        pb9.compat(),
+        pa0.compat(),
+        delay.compat(),
+        &CONFIG_RADIO,
+    )
+    .unwrap(); // should handle error
+
+    (tx, rx, lora)
+}
+
+#[cfg(feature = "stm32f1xx")] //  eg blue pill stm32f103
+use stm32f1xx_hal::{
+    delay::Delay,
+    device::USART2,
+    pac::Peripherals,
+    prelude::*,
+    serial::{Config, Rx, Serial, Tx}, //, StopBits
+    spi::{Error, Spi},
+};
+
+#[cfg(feature = "stm32f1xx")]
```

Comment:
okay so this platform-specific setup logic could be pulled out into a common file shared by each of the examples, which would get rid of a few LoC? and you can combine the output impl constraints to `Transmit` and `Request`, then the main file becomes just the example logic.

Reply:
Thanks for the rebase tips.

I considered doing this but could not figure out how to do something like include/source and then return. (One more thing that I know how to do in several languages but have not figured out in rust.) I didn't want to put it as a module in  src/  because that makes it less easily accessible to someone reading the example.

I do think it would be a good idea if it can be done as a separate file in the examples/ directory.

Reply:
I've looked more at doing this. The Cargo Book  suggests a bigger example can be split into files at https://doc.rust-lang.org/edition-guide/rust-2018/cargo-and-crates-io/multi-file-examples.html#multi-file-examples but does not explain much. The blog http://xion.io/post/code/rust-examples.html gives a bit more detail. It does not appear to be a widely used capability, and I think may be somewhat experimental. 

I played with it by splitting an example in two in a subdirectory as indicated in the links above, with the setup functions in one file. The parts seem to compile separately rather than just pulling `setup.rs` into `main.rs` and compiling it all together. (A very `Rust` way of doing it I guess.) Because of separate compiles the `use` statements need to be in the correct files, and in some cases duplicated it seems.  I made some progress on this but then all the `.compat()` stuff seems to get lost. No matter which file I put `use embedded_hal_compat` the other file complained, and putting it in both also did not work.

The separate files would also not be interesting if the setup for all examples could not be the same file for all examples. So I tried moving  it above the example directory. That causes the additional problem that the module is not found. In linux it might be possible to workaround using soft links, but I seem to recall that Windows does not like that much. 

At this point, my preference would be to leave the examples as they are, or add a simple comment that a user can just ignore any setup for a platform that they are not using.

Reply:
I have played with this more and the two reasonable solutions seem to be:
- Leave the (mostly duplicated) common setup in the two examples. This makes the two example files long but the structure is easy to follow. For anyone interested in a specific MCU the relevant part is small and fairly obvious.
- Move the common setup to `src/`. (I have not yet explored this carefully.) It makes the two example files short but an important part of the example is in a file in `src/` rather than in the example file itself. For me, this complicates reading the example. It does have the advantage that it illustrates how setup functions might be used in a crate. It has a disadvantage that it limits the possibility of putting more in `[dev-dependencies]`. I don't yet fully understand release implications of  `[dev-dependencies]` vs `[dependencies]` with `optional = true`.

Please let me know your preference.

Reply:
Sorry, three example files, `lora_spi_send.rs`, `lora_spi_receive.rs`, and` lora_spi_gps.rs`. Each uses slightly different traits and has two MCU setups.